marshmallow_dataclass/__init__.py

```diff
@@ -996,6 +1023,66 @@ def _get_field_default(field: dataclasses.Field):
     return field.default
 
 
+def is_generic_alias_of_dataclass(clazz: type) -> bool:
+    """
+    Check if given class is a generic alias of a dataclass, if the dataclass is
+    defined as `class A(Generic[T])`, this method will return true if `A[int]` is passed
+    """
+    is_generic = is_generic_type(clazz)
+    type_arguments = typing_extensions.get_args(clazz)
+    origin_class = typing_extensions.get_origin(clazz)
+    return (
+        is_generic
+        and len(type_arguments) > 0
+        and dataclasses.is_dataclass(origin_class)
+    )
+
+
+def _get_type_hints(
+    obj,
+    schema_ctx: _SchemaContext,
+):
+    if sys.version_info >= (3, 9):
+        type_hints = get_type_hints(
+            obj,
+            globalns=schema_ctx.globalns,
+            localns=schema_ctx.localns,
+            include_extras=True,
+        )
+    else:
+        type_hints = get_type_hints(
+            obj, globalns=schema_ctx.globalns, localns=schema_ctx.localns
+        )
+
+    return type_hints
+
+
+def _get_generic_type_hints(
+    obj,
+    schema_ctx: _SchemaContext,
+) -> type:
+    """typing.get_type_hints doesn't work with generic aliasses. But this 'hack' works."""
```

Comment:
Hmm, the forward refs have already been resolved. This really exists purely to get the typehint for generics.
maybe `_get_type_hint_of_generic_object` ?


```python
import typing

T = typing.TypeVar('T')

class A(Generic[T]):
  a: T

class B(A[int]):
  pass

print(typing.get_type_hints(B))
print(typing.get_type_hints(A[int]))

>=========================
{'a': ~T}
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[51], [line 12](vscode-notebook-cell:?execution_count=51&line=12)
      [9](vscode-notebook-cell:?execution_count=51&line=9)   pass
     [11](vscode-notebook-cell:?execution_count=51&line=11) print(typing.get_type_hints(B))
---> [12](vscode-notebook-cell:?execution_count=51&line=12) print(typing.get_type_hints(A[int]))

File ~\.pyenv\pyenv-win\versions\3.11.3\Lib\typing.py:2347, in get_type_hints(obj, globalns, localns, include_extras)
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2345)         return {}
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2346)     else:
-> ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2347)         raise TypeError('{!r} is not a module, class, method, '
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2348)                         'or function.'.format(obj))
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2349) hints = dict(hints)
   ~/.pyenv/pyenv-win/versions/3.11.3/Lib/typing.py:2350) for name, value in hints.items():

TypeError: __main__.A[int] is not a module, class, method, or function.
```

```python
def _get_generic_type_hints(obj) -> type:
    """typing.get_type_hints doesn't work with generic aliases. But this 'hack' works."""

    class X:
        x: obj  # type: ignore[name-defined]

    return typing.get_type_hints(X)['x']

print(_get_generic_type_hints(B))
print(_get_generic_type_hints(A[int]))

>=========================
<class '__main__.B'>
__main__.A[int]
```

Reply:
In your last examples, `_get_generic_type_hints` is just the identity, right?

```py
>>> _get_generic_type_hints(B) is B
True
>>> _get_generic_type_hints(A[int]) is A[int]
True
```

But, the value of `_get_generic_type_hints` comes in resolving forward references:
```
>>> _get_generic_type_hints(A["int"])
__main__.A[int]
>>> _get_generic_type_hints(A["int"]) is A[int]
False
```

Reply:
Maybe what is really wanted is (untested)
```py
def _get_generic_type_hints(generic_alias):
    class X(generic_alias): pass

    return typing.get_type_hints(X)
```

Or, in one line:
```
    return typing.get_type_hints(types.new_class("_", bases=(generic_alias,)))
```

Reply:
Wow, good catch! You're absolutely right.

1. This just worked 🤦‍♂️ I clearly spend too long trying to make it work to see the obvious solution.
```python
(
    type_hints[field.name]
    if not is_generic_type(clazz)
    else field.type
)
```

2. But didn't work for `A["int"]` and no tests caught that yet. So I'll add a tests and rename the function as per your recommendation.

Reply:
And, if that works, maybe just roll this into `_get_type_hints` so that there's a single `_get_type_hints` function that can be used for any dataclass, be it generic or not.

Reply:
> ```python
> def _get_generic_type_hints(generic_alias):
>     class X(generic_alias): pass
>
>     return typing.get_type_hints(X)
> ```

Does not work.
```python
print(_get_generic_type_hints(A[int]))

>=====================
{'a': ~T}
```

Instead of `{'a': __main__.A[int]}`

Reply:
> Does not work.

That's too bad.

I still think that refactoring so that `_get_type_hints` will work with generic aliases of dataclasses as well as plain dataclasses is probably worth it.  It will move all the `is_generic_type` special-casing out of 

https://github.com/lovasoa/marshmallow_dataclass/blob/7ac088d6086c5e4e2313bced99f1454204834563/marshmallow_dataclass/__init__.py#L583-L603

Reply:
I don't disagree but I've already sunk too many hours into this particular problem. If it is even possible, then I don't have the knowledge to take this further.

Reply:
@dairiki I found a way around it by resolving the forward refs when we get the dataclass fields. We now no longer call `get_type_hints` at all anymore as it's relevant code is internalized. 

We were already looping over the mro and fields just like `get_type_hints`.